size_test.go

```diff
@@ -82,6 +82,78 @@ func TestHumanSize(t *testing.T) {
+	assertEquals(t, "1e+04YB", HumanSizeWithPrecision(float64(10000000000000*PB), 4))
+
+	assertEquals(t, "1kB", HumanSizeWithPrecision(1000, 3))
+	assertEquals(t, "1.02kB", HumanSizeWithPrecision(1024, 3))
+	assertEquals(t, "1MB", HumanSizeWithPrecision(1000000, 3))
+	assertEquals(t, "1.05MB", HumanSizeWithPrecision(1048576, 3))
+	assertEquals(t, "2MB", HumanSizeWithPrecision(2*MB, 3))
+	assertEquals(t, "3.42GB", HumanSizeWithPrecision(float64(3.42*GB), 3))
+	assertEquals(t, "5.37TB", HumanSizeWithPrecision(float64(5.372*TB), 3))
+	assertEquals(t, "2.22PB", HumanSizeWithPrecision(float64(2.22*PB), 3))
+	assertEquals(t, "1e+04YB", HumanSizeWithPrecision(float64(10000000000000*PB), 3))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(999.5*MB), 3))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(999.9*MB), 3))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(1000*MB), 3))
+
+	assertEquals(t, "1kB", HumanSizeWithPrecision(1000, 2))
+	assertEquals(t, "1kB", HumanSizeWithPrecision(1024, 2))
+	assertEquals(t, "1MB", HumanSizeWithPrecision(1000000, 2))
+	assertEquals(t, "1MB", HumanSizeWithPrecision(1048576, 2))
+	assertEquals(t, "2MB", HumanSizeWithPrecision(2*MB, 2))
+	assertEquals(t, "3.4GB", HumanSizeWithPrecision(float64(3.42*GB), 2))
+	assertEquals(t, "5.4TB", HumanSizeWithPrecision(float64(5.372*TB), 2))
+	assertEquals(t, "2.2PB", HumanSizeWithPrecision(float64(2.22*PB), 2))
+	assertEquals(t, "1e+04YB", HumanSizeWithPrecision(float64(10000000000000*PB), 2))
+	assertEquals(t, "0.1GB", HumanSizeWithPrecision(float64(99.5*MB), 2))
+	assertEquals(t, "0.1GB", HumanSizeWithPrecision(float64(99.9*MB), 2))
+	assertEquals(t, "0.1GB", HumanSizeWithPrecision(float64(100*MB), 2))
+	assertEquals(t, "0.9GB", HumanSizeWithPrecision(float64(900*MB), 2))
+	assertEquals(t, "0.95GB", HumanSizeWithPrecision(float64(950*MB), 2))
+	assertEquals(t, "0.96GB", HumanSizeWithPrecision(float64(960*MB), 2))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(996*MB), 2))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(999.5*MB), 2))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(999.9*MB), 2))
+	assertEquals(t, "1GB", HumanSizeWithPrecision(float64(1000*MB), 2))
+
+	// HumanSizeWithPrecision does not work with precision 1 very well
+	assertEquals(t, "1kB", HumanSizeWithPrecision(1000, 1))
+	assertEquals(t, "1kB", HumanSizeWithPrecision(1024, 1))
+	assertEquals(t, "1MB", HumanSizeWithPrecision(1000000, 1))
+	assertEquals(t, "1MB", HumanSizeWithPrecision(1048576, 1))
+	assertEquals(t, "2MB", HumanSizeWithPrecision(2*MB, 1))
+	assertEquals(t, "3GB", HumanSizeWithPrecision(float64(3.42*GB), 1))
+	assertEquals(t, "6TB", HumanSizeWithPrecision(float64(5.972*TB), 1))
+	assertEquals(t, "2PB", HumanSizeWithPrecision(float64(2.22*PB), 1))
+	assertEquals(t, "1e+04YB", HumanSizeWithPrecision(float64(10000000000000*PB), 1))
+	assertEquals(t, "0.009GB", HumanSizeWithPrecision(float64(9.5*MB), 1))
```

Comment:
I'm a bit confused, what is the precision int supposed to signify? I can't understand how "0.009GB" is a correct expectation for 9.5MB with a precision of 1

Reply:
I'd have to look as well; not all of this was ever properly "documented", and some of the behavior most definitely got into the codebase "organically" (parts of this code at least originated from the moby/moby ("docker engine") codebase.

I know that long (long) time ago I opened a ticket related to output that confused me, which was because different number of decimals were used in the output which, combined with confusion about decimal separators (`.` vs `,` differing between locales) at least confused "me"; https://github.com/moby/moby/issues/9977#issue-53772370

Reply:
Searching within the `docker` and `moby` orgs here on GitHub:

- `moby/containerd` only seems to use `CustomSize()` in one spot;
-  In the `docker` org we seem to be using `HumanSizeWithPrecision()` in:
- - `cli`;
- - `compose`;
- - `docker-ce`;
- - `registry-cli`
- and `CustomSize()` only in the `cs-libnetwork` repo. 

Not a ton of occurrences, so I'd be happy if we could come together on a better definition for this func and make the required changes wherever they may be needed. 

If we agree to continue using the word "precision" to represent the total number of digits in the size representation, as it seems to currently be, instead of the number of digits after the decimal point (which I think would be a much better way of both defining and using the func) then I think we should adjust how the function works so that the output is at least consistent for all precision values used.

All values should be approximated anyway, so if we need to approximate 9.5*MB with a "precision" of 1, then I think the result should be the closest representation using a single digit, so for example 9MB instead of 0.009GB.

From what I've managed to find searching around, we never use precision values of 1 or 2, so I wouldn't worry too much about the "validity" of the output given that it's an approximation. We can always write better documentation on the functions so things are a bit clearer for the users, or outright make 3 the minimum precision value.

eg To be consistent, I think these tests should pass
```
assertEquals(t, "9MB", HumanSizeWithPrecision(float64(9.5*MB), 1))
assertEquals(t, "9MB", HumanSizeWithPrecision(float64(9.9*MB), 1))
assertEquals(t, "9MB", HumanSizeWithPrecision(float64(100*MB), 1))
assertEquals(t, "1GB", HumanSizeWithPrecision(float64(500*MB), 1))
assertEquals(t, "1GB", HumanSizeWithPrecision(float64(900*MB), 1))
assertEquals(t, "1GB", HumanSizeWithPrecision(float64(1499*MB), 1))
```

My personal preference would be to fix the definition and use actual number of digits after the decimal point as the meaning of `precision`, maybe by writing a new func and adjusting existing code that uses the old func to use the new func before removing it entirely.

Reply:
Hi @krissetto, I'm looking forward to a final decision. Thanks!

Reply:
Hey @kmirzavaziri! 

I don't think I'll be able to give a final decision here as I'm also new to Docker and I don't know how we manage these kinds of situations, but I'll leave my personal opinion and some observations here, waiting to be corrected by the powers that be ;)

**TL;DR**  
I would think about creating a new function with more consistent behavior and naming,  migrating our internal codebase to use it and marking the old func as deprecated somehow, but leaving it available in case any external codebases are using it.  I don't think the "width" of the response should be part of the func signature, but only the precision (and maybe an optional maximum desired unit of size? eg `TB`). Handling the actual display size of the output seems to me to be a responsibility of whatever client decides to consume this library, and not something we should be dealing with here.
 
---

_Some observations_

I don't really like how we're using the concept of precision currently, as it goes against most of the stdlib fmt use-cases (which consider both width - min number of unicode runes - and precision - chars to the right of the decimal point). What we're actually using is the concept of "significant digits", which Go seems to handle in a peculiar way when using the `%.*g` syntax.   

Take a look at the following test case we have:
`assertEquals(t, "0.009GB", WowWithPrecision(float64(9.5*MB), 1))`

Here the result ends up being 0.009GB, as 9 is the precision of one (as in 1 significant digit). This is still somewhat ok as it's still human-readable, but I think we should adjust our logic to leave the representation as 9.5MB, which is shorter in width, more human-readable, and actually has a "precision" of 1 (the kind of precision we like).

Checking this test case instead:
`assertEquals(t, "1e+04YB", HumanSizeWithPrecision(float64(10000000000000*PB), 1))`

I think we have more of a proble. Using the `%.*g` syntax, Go is converting the output to scientific notation in order to keep 1 significant digit, the 4 at the very end. My guess is that 1e+0.. is not considered as significant because it's part of the standard scientific notation, but I'm sure most-all of us would agree that it's not exactly human readable.  `∞` would probably be better after a certain file size 😂 

If we were to use `%.*f` instead, the output would become `10000.0YB`. 1 digit of (in this case unnecessary) precision, but a relatively "wide" total number since we currently max out our units at YB, and thus cannot divide any further.

In the first test example, using `9.5MB`, the func would return 0.0GB using `%.*f` and a precision of 1, since we are still doing one too many divisions and don't have enough precision available to represent the value at all.